odxtools/codec.py

```diff
@@ -0,0 +1,212 @@
+    the wire implements this API.
+    """
+
+    @property
+    def short_name(self) -> str:
+        return ""
+
+    def encode_into_pdu(self, physical_value: Optional[ParameterValue],
+                        encode_state: EncodeState) -> None:
+        ...
+
+    def decode_from_pdu(self, decode_state: DecodeState) -> ParameterValue:
+        ...
+
+    def get_static_bit_length(self) -> Optional[int]:
+        ...
+
+
+@runtime_checkable
+class CompositeCodec(Codec, typing.Protocol):
+    """Any object which can be en- or decoded to be transferred over
+    the wire which is composed of multiple parameter implements this
+    API.
+
+    """
+
+    @property
+    def parameters(self) -> List["Parameter"]:
+        return []
+
+    @property
+    def required_parameters(self) -> List["Parameter"]:
+        return []
+
+    @property
+    def free_parameters(self) -> List["Parameter"]:
+        return []
+
+
+# some helper functions useful for composite codec objects
+def composite_codec_get_static_bit_length(codec: CompositeCodec) -> Optional[int]:
+    """Compute the length of a composite codec object in bits
+
+    This is basically the sum of the lengths of all parameters. If the
+    length of any parameter can only determined at runtime, `None` is
+    returned.
+    """
+
+    cursor = 0
+    byte_length = 0
+    for param in codec.parameters:
+        param_bit_length = param.get_static_bit_length()
+        if param_bit_length is None:
+            # We were not able to calculate a static bit length
+            return None
+        elif param.byte_position is not None:
+            cursor = param.byte_position
+
+        cursor += ((param.bit_position or 0) + param_bit_length + 7) // 8
+        byte_length = max(byte_length, cursor)
+
+    # Round up to account for padding bits (all structures are
+    # byte aligned)
+    return byte_length * 8
+
+
+def composite_codec_get_required_parameters(codec: CompositeCodec) -> List["Parameter"]:
+    """Return the list of parameters which are required to be
+    specified for encoding the structure
+
+    I.e., all free parameters that do not exhibit a default value.
+    """
+    return [p for p in codec.parameters if p.is_required]
+
+
+def composite_codec_get_free_parameters(codec: CompositeCodec) -> List["Parameter"]:
+    """Return the list of parameters which can be freely specified by
+    the user when encoding the structure
+
+    This means all required parameters plus parameters that can be
+    omitted because they specify a default.
+    """
+    return [p for p in codec.parameters if p.is_settable]
+
+
+def composite_codec_get_coded_const_prefix(codec: CompositeCodec,
+                                           request_prefix: bytes = b'') -> bytes:
+    from .parameters.codedconstparameter import CodedConstParameter
+    from .parameters.matchingrequestparameter import MatchingRequestParameter
+    from .parameters.physicalconstantparameter import PhysicalConstantParameter
```

Comment:
why is the import here?

Reply:
because it is needed to compute the static bit size of composite codec objects...

Reply:
because in the original version of the PR it would have lead to cyclic imports. That said, if we move these imports to the top we can also import `Parameter` (i.e. forget about my comment above.)

Reply:
fixed.